File: 07/OptimalPageReplacement.java
import java.util.*;

public class OptimalPageReplacement {
    public static void optimal(int[] pages, int numFrames) {
        int pageFaults = 0;
        int[] frames = new int[numFrames];
        Arrays.fill(frames, -1);  // Initialize frames to -1 (empty)

        for (int i = 0; i < pages.length; i++) {
            int page = pages[i];
            if (!contains(frames, page)) {
                pageFaults++;
                if (getEmptySlot(frames) != -1) {
                    frames[getEmptySlot(frames)] = page;  // Place page in empty slot
                } else {
                    int indexToReplace = getOptimalPageToReplace(frames, pages, i);
                    frames[indexToReplace] = page;  // Replace the optimal page
                }
            }

            // Print current frame contents
            System.out.print("Frames: ");
            for (int j : frames) {
                System.out.print(j + " ");
            }
            System.out.println();
        }

        System.out.println("Total Page Faults: " + pageFaults);
    }

    private static boolean contains(int[] frames, int page) {
        for (int frame : frames) {
            if (frame == page) {
                return true;
            }
        }
        return false;
    }

    private static int getEmptySlot(int[] frames) {
        for (int i = 0; i < frames.length; i++) {
            if (frames[i] == -1) {
                return i;
            }
        }
        return -1;
    }

    private static int getOptimalPageToReplace(int[] frames, int[] pages, int currentIndex) {
        int farthest = currentIndex;
        int pageToReplace = -1;
        for (int i = 0; i < frames.length; i++) {
            int j;
            for (j = currentIndex + 1; j < pages.length; j++) {
                if (frames[i] == pages[j]) {
                    break;
                }
            }
            if (j == pages.length) {
                return i;  // The page is not needed anymore
            }
            if (j > farthest) {
                farthest = j;
                pageToReplace = i;
            }
        }
        return pageToReplace;
    }

    public static void main(String[] args) {
        int[] pages = {7, 0, 1, 2, 0, 3, 0, 4, 2, 3};
        int numFrames = 3;
        optimal(pages, numFrames);
    }
}

File: 07/OptimalPageReplacement.py
def optimal(pages, num_frames):
    frames = [-1] * num_frames  # Initialize frames to -1 (empty)
    page_faults = 0

    def contains(frames, page):
        return page in frames

    def get_empty_slot(frames):
        try:
            return frames.index(-1)
        except ValueError:
            return -1

    def get_optimal_page_to_replace(frames, pages, current_index):
        farthest = current_index
        page_to_replace = -1
        for i in range(len(frames)):
            for j in range(current_index + 1, len(pages)):
                if frames[i] == pages[j]:
                    break
            if j == len(pages):
                return i  # The page is not needed anymore
            if j > farthest:
                farthest = j
                page_to_replace = i
        return page_to_replace

    for i in range(len(pages)):
        page = pages[i]
        if not contains(frames, page):
            page_faults += 1
            empty_slot = get_empty_slot(frames)
            if empty_slot != -1:
                frames[empty_slot] = page
            else:
                replace_index = get_optimal_page_to_replace(frames, pages, i)
                frames[replace_index] = page

        # Print current frame contents
        print("Frames:", frames)

    print("Total Page Faults:", page_faults)

# Test the function
pages = [7, 0, 1, 2, 0, 3, 0, 4, 2, 3]
num_frames = 3
optimal(pages, num_frames)
